File: Iteration2/code/CSE3063GRP12/src/features/main_menu/MenuController.java
package features.main_menu;

import java.util.ArrayList;

import core.enums.Menu;
import core.database.abstracts.DatabaseManager;
import core.enums.AdvisorMenu;
import core.enums.StudentMenu;

import core.enums.UserType;
import core.exceptions.UnexpectedInputException;
import core.general_providers.AppConstant;
import core.general_providers.InstanceManager;
import core.general_providers.SessionController;
import core.general_providers.TerminalManager;
import core.models.abstracts.User;
import core.models.concretes.NotificationResponse;
import core.repositories.NotificationRepositories;

public class MenuController {
    private DatabaseManager databaseManager;
    private NotificationRepositories notificationRepositories;
    private MenuView menuView;

    // Constructor
    public MenuController() {
        this.menuView = new MenuView();
        this.notificationRepositories = new NotificationRepositories();
        handleMenu();
    }

    // handle menu
    public void handleMenu() {
        // Get menu items
        ArrayList<String> menuItems = getMenuItems();
        // Show menu list
        menuView.showMenuList(menuItems);
        // Get menu selection
        String menuSelection = null;

        try{
            menuSelection = getMenuSelection();
            if(Integer.parseInt(menuSelection) < 1 || Integer.parseInt(menuSelection) > menuItems.size())
                throw new UnexpectedInputException();
        }catch(UnexpectedInputException e){
            menuView.showErrorMessage(e);
            handleMenu();
        }
        
        // Convert menu from string
        Menu menu = convertEnum(menuSelection);
        // Navigate to module
        navigateToModule(menu);
    }

    // Method to get menu items
    public ArrayList<String> getMenuItems() {
        ArrayList<String> menuItems = new ArrayList<String>();
        UserType userType = SessionController.getInstance().getUserType();
        //UserType userType = UserType.Student; //change here with upper line
        
        boolean containNew= getNotification(userType);
        switch (userType) {
            case Student:
                for (StudentMenu item : StudentMenu.values()) {
                    if( item.isNotification()&& containNew){
                        menuItems.add("\033[1m" +"*"+ item.getItemMessage() +"*"+ "\033[0m ");
                       }else{
                        menuItems.add(item.getItemMessage());
                       }
                }
                break;
            case Advisor:
                for (AdvisorMenu item : AdvisorMenu.values()) {
                    System.out.println(containNew);
                    if( item.isNotification()&& containNew){
                        menuItems.add("\033[1m" +"*"+ item.getItemMessage() +"*"+ "\033[0m ");
                       }else{
                        menuItems.add(item.getItemMessage());
                       }}
                break;
        }
        return menuItems;
    }

    public boolean getNotification(UserType userType){
        User user = SessionController.getInstance().getCurrentUser();
        NotificationResponse notificationResponse = null;
        try{
            if (userType == UserType.Student) {
                notificationResponse = notificationRepositories.getNotification(user.getUserName());
            } else if (userType == UserType.Advisor) {
                notificationResponse = notificationRepositories.getNotification(user.getUserName());
            }
            if (notificationResponse == null) {
                // throw new UserNotFoundException();
                return false;
            }else{
                return notificationResponse.isContainsNew();
            }
        } catch (Exception e) {
            e.printStackTrace();
            // throw new UserNotFoundException();
        }

        return true;
    }
    // navigate to module
    public void navigateToModule(Menu menu) {
        menu.navigate();
    }
    
    // get menu selection from user
    public String getMenuSelection() throws UnexpectedInputException{ //this will return UnexpectedInputException
        menuView.showPromptMessage();
        String selection = TerminalManager.getInstance().read();
        // Check if selection is a number
        if(selection.matches("^\\d+$") == false){
            throw new UnexpectedInputException();
        } 
        return selection; 
    }

    // convert menu from string
    public Menu convertEnum(String menuString) {
        // convert string to Menu enum
        Menu menu;
        if (SessionController.getInstance().getUserType() == UserType.Advisor)
            menu = AdvisorMenu.values()[Integer.parseInt(menuString) - 1];
        else
            menu = StudentMenu.values()[Integer.parseInt(menuString) - 1];
        return menu;
    }
}

File: Iteration3/code/CSE3063GRP12/src/features/main_menu/MenuController.py
from features.main_menu.MenuView import MenuView
from core.general_providers.SessionController import SessionController
from core.repositories.NotificationRepository import NotificationRepository
from core.enums.StudentMenu import StudentMenu
from core.enums.AdvisorMenu import AdvisorMenu
from core.enums.UserType import UserType
from core.exceptions.UnexpectedInputException import UnexpectedInputException

class MenuController:
    def __init__(self):
        self.menu_view = MenuView()
        self.notification_repository = NotificationRepository()
        self.handle_menu()

    def handle_menu(self):
        # Get menu items
        menu_items = self.get_menu_items()
        # Show menu list
        self.menu_view.show_menu_list(menu_items)
        # Get menu selection
        menu_selection = None

        try:
            menu_selection = self.get_menu_selection()
            if int(menu_selection) < 1 or int(menu_selection) > len(menu_items):
                raise UnexpectedInputException()
            # Convert menu from string
            menu = self.convert_enum(menu_selection)
            # Navigate to module
            self.navigate_to_module(menu)
        except UnexpectedInputException as e:
            self.menu_view.show_error_message(e)
            self.handle_menu()



    def get_menu_items(self):
        menuItems = []
        userType = SessionController.getInstance().get_user_type()

        containNew = self.get_notification(userType)
        if userType == UserType.Student:
            for item in StudentMenu:
                if item.is_notification() and containNew:
                    menuItems.append(f"*{item.get_item_message()}*")
                else:
                    menuItems.append(item.get_item_message())
        elif userType == UserType.Advisor:
            for item in AdvisorMenu:
                if item.is_notification() and containNew:
                    menuItems.append(f"*{item.get_item_message()}*")
                else:
                    menuItems.append(item.get_item_message())

        return menuItems  

    def get_notification(self, userType):
        user = SessionController.getInstance().get_current_user()
        notificationResponse = None
        try: # bu try hata veriyor AttributeError('STUDENT')
            if userType == UserType.Student or userType == UserType.Advisor:
                notificationResponse = self.notification_repository.get_notification(user.userName) #sorun burda

            if notificationResponse is None:
                return False
            else:
                return notificationResponse.containsNew
        except Exception as e:
            print(e)

        return True

    def navigate_to_module(self, menu):
        menu.navigate()

    def get_menu_selection(self):
        selection = input("Please enter your selection: ")

        if not selection.isdigit():
            raise UnexpectedInputException()
        return selection
    
    def convert_enum(self, menu_selection):
        menu = None
        user_type = SessionController.getInstance().get_user_type()


        if user_type == UserType.Student:
            menu_list = list(StudentMenu)
            menu = menu_list[int(menu_selection)-1]
        elif user_type == UserType.Advisor:
            menu_list = list(AdvisorMenu)
            menu = menu_list[int(menu_selection)-1]

        return menu
    

